### PR TITLE
Help text tests

### PR DESCRIPTION
**Pull request by drewburlingame**
@bilal-fazlani, this has the test for #74.  I don't like capturing the help text this way.  It will have race conditions with other tests when run in parellel.    Since help is output to the console, I don't know of a good way to isolate the output for the running test.  Do you have any suggestions?

### CODE REVIEW

**Comment by bilal-fazlani** [comment]
Trying to capture the output ourselves and then doing string comparisons will most like increase testing effort by a lot.

I think we should write these tests at a higher level, leveraging a cli testing tool. I found an example here. https://spin.atomicobject.com/2016/01/11/command-line-interface-testing-tools/

I am sure there will be other tools as well.

File: CommandDotNet.Tests/HelpTests/UsageTests.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            _output.Add(e);
+        }
+
+        [Fact]
+        public void UsageShouldPrintSubcommandsInCorrectOrder()
+        {
+            new AppRunner<HelpApp>().Run("SubHelpCommand", "Print", "--help");
+            var helpText = _output.First().Split(new[] {"\r\n", "\r", "\n"}, StringSplitOptions.RemoveEmptyEntries);
+            var usage = helpText.Single(o => o.StartsWith("Usage")).Substring("Usage: ".Length);
```
Comment (bilal-fazlani) [inline CommandDotNet.Tests/HelpTests/UsageTests.cs]: ummm., this looks too fragile.

Comment (drewburlingame) [reply CommandDotNet.Tests/HelpTests/UsageTests.cs]: I agree to an extent.  Parsing strings is generally fragile, and so is comparing on strings.  The well-defined format can let us create a parser to assert subsections of the output.  If the format changes, the parser can be changed to fix all tests.  If we compare on the full output, every test will have to be updated if the format changes.

**Comment by drewburlingame** [comment]
> Trying to capture the output ourselves and then doing string comparisons will most like increase testing effort by a lot.
> 
> I think we should write these tests at a higher level, leveraging a cli testing tool. I found an example here. https://spin.atomicobject.com/2016/01/11/command-line-interface-testing-tools/
> 
> I am sure there will be other tools as well.

I was thinking along the same lines.  I remembered you posted about https://github.com/simonmichael/shelltestrunner in another issue.  I like the idea of it.  I don't believe most of these tools are great options because they're either not easily portable or grokable.   Bats & Empty require Bash.  Arubas requires Ruby.  Expect requires Tcl which requires an ActiveState download for Windows.  I wonder if it would be more work to maintain these and harder to for .net devs to contribute.

An alternative is to spawn a new process and capture the output that way.  It avoids the static Console.Out reference so it would support mullti-threaded tests, supports multi-platform and multi-console testing, and .net devs can use the same tools they're familiar with to contribute.

What do you think?

**Comment by bilal-fazlani** [comment]
Your points about these tools and tests are valid. They are hard to maintain. And that’s is why they should be very less in number. Our intent with these tests should not be to test every use case but just catch high level bugs such as recent one where help text got reversed. By making sure that there are just a couple of such tests, we ensure that overhead of maintaining these tests is less.

These should also not sit in the same stage as unit tests in our CI pipeline.

image

To me, these are more like snapshot tests, https://jestjs.io/docs/en/snapshot-testing 
Or UI smoke tests. 

Let me try a test case with shelltestrunner and then let’s compare. What say?

**Comment by bilal-fazlani** [comment]
btw, I tried shelltest runner... and encountered some issues. It did not give me options to ignore "\r" , "\n" and whitespaces. So got a lot of false negatives.  Will spend some more time on this .. probably will explore another tool

**Pull request closed without merging**